### PR TITLE
Search interface

### LINKED ISSUE #14: Search implementation

### PR DESCRIPTION
**Pull request by JuniorJPDJ**
First interface, to be verified by FUMR/tidal-async impl.
resolves #14 

Example impl: FUMR/tidal-async#74

### CODE REVIEW

File: music_service_async_interface/__init__.py [hunk 1]
```diff
@@ -254,6 +270,13 @@ def cover(self) -> Optional["Cover"]:
         ...
 
 
+class Searchable(Object, ABC):
+    """Generic class to be subclassed by any searchable :class:`Object`
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: s/subclassed/extended/

File: music_service_async_interface/__init__.py [hunk 2]
```diff
@@ -254,6 +270,13 @@ def cover(self) -> Optional["Cover"]:
...
+    Eg. if music service is abled to search for tracks and
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: s/Eg/For example/
s/d//

File: music_service_async_interface/__init__.py [hunk 3]
```diff
@@ -254,6 +270,13 @@ def cover(self) -> Optional["Cover"]:
...
+    playlists, `Track` and `Playlist` classes should subclass this class.
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: s/subclass/extend/

File: music_service_async_interface/__init__.py [hunk 4]
```diff
@@ -197,6 +197,22 @@ async def parse_urls(self, long_string: str) -> AsyncGenerator["Object", None]:
                 except InvalidURL:
                     pass
 
+    @abstractmethod
+    async def search(
+        self, query, types: Optional[Union[Type["Searchable"], List[Type["Searchable"]]]] = None
+    ) -> AsyncGenerator["Searchable", None]:
+        """Searches music service for specific type(s) of :class:`Searchable` objects using `query`
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: either `types of objects` or `type of object`
s/query/query string/

if it refers to the Searchable class, then phrase it like: `Searches music service for specific Searchable types using....`

File: music_service_async_interface/__init__.py [hunk 5]
```diff
@@ -197,6 +197,22 @@ async def parse_urls(self, long_string: str) -> AsyncGenerator["Object", None]:
...
+        Accepts single :class:`Searchable` object or list of them.
+
+        :param query: search string/phrase
+        :param types: :class:`Searchable` object or list of objects, filters search results to this type,
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: Is it an object or a class type

File: music_service_async_interface/__init__.py [hunk 6]
```diff
@@ -197,6 +197,22 @@ async def parse_urls(self, long_string: str) -> AsyncGenerator["Object", None]:
...
+        eg. if you want to search for tracks use `sess.search('text', Track)`
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: s/eg./e. g./

**Review by wvffle — CHANGES_REQUESTED** [review]

File: music_service_async_interface/__init__.py (hunk 5 above)
Comment (JuniorJPDJ) [reply music_service_async_interface/__init__.py]: oh fuck.

File: music_service_async_interface/__init__.py [hunk 7]
```diff
@@ -197,6 +201,24 @@ async def parse_urls(self, long_string: str) -> AsyncGenerator["Object", None]:
...
+        self, query: str, types: Optional[Union[Type["Searchable"], List[Type["Searchable"]]]] = None, limit: int = 10
+    ) -> AsyncGenerator["Searchable", None]:
+        """Searches music service for specific :class:`Searchable` types using `query`
+        Accepts single :class:`Searchable` type, list of them or allows to entirely skip `types` parameter.
+
+        :param query: search string/phrase
+        :param types: :class:`Searchable` types or list of types, filters search results to this type,
+        e.g. if you want to search for tracks use `sess.search('text', Track)`
+        or if you want to search for albums and playlists use `sess.search('text', [Album, Playlist])`.
+        Also allows to search for all possible objects, then this parameter is left unset or set to None/empty list.
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: Same as in the tidal implementation

File: music_service_async_interface/__init__.py [hunk 8]
```diff
@@ -254,6 +276,13 @@ def cover(self) -> Optional["Cover"]:
         ...
 
 
+class Searchable(Object, ABC):
+    """Generic class to be extended by any searchable :class:`Object`
+    For example if music service is abled to search for tracks and
```
Comment (wvffle) [inline music_service_async_interface/__init__.py]: abled

**Review by wvffle — CHANGES_REQUESTED** [review]

**Comment by JuniorJPDJ** [comment]
[T. Hanks](https://en.wikipedia.org/wiki/Tom_Hanks) for review @wvffle, is it fixed now?

**Comment by JuniorJPDJ** [comment]
Just rebasing.

**Review by wvffle — APPROVED** [review]

**Pull request merged**